### PR TITLE
[SPARK-36649][SQL] Support `Trigger.AvailableNow` on Kafka data source

### PR DESCRIPTION
**Pull request by jerrypeng**
### What changes were proposed in this pull request?
To add support to using the AvailableNow trigger for the KafkaSource

### Why are the changes needed?
To allow users to use the KafkaSource the AvailableNow trigger.  The AvailableNow trigger allows the all the data currently available in the source to be processed in multiple micro batches.

### Does this PR introduce _any_ user-facing change?

Yes, allows users to using the KafkaSource to use the AvailableNow trigger

### How was this patch tested?

Added a test for this.

### CODE REVIEW

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala [hunk 1]
```diff
@@ -130,7 +134,7 @@ private[kafka010] class KafkaSource(
     } else if (minOffsetPerTrigger.isDefined) {
       ReadLimit.minRows(minOffsetPerTrigger.get, maxTriggerDelayMs)
     } else {
-      maxOffsetsPerTrigger.map(ReadLimit.maxRows).getOrElse(super.getDefaultReadLimit)
+      maxOffsetsPerTrigger.map(ReadLimit.maxRows).getOrElse(ReadLimit.allAvailable() )
```
Comment (jerrypeng) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala:137]: Had to make this change because of scala compiler bug that I encountered:

https://github.com/scala/bug/issues/12523

Comment (HeartSaVioR) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala:137]: nit: no space before )

Comment (jerrypeng) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala]: fixed

**Comment by jerrypeng** [comment]
> Welcome to Spark community and thanks for the contribution @jerrypeng !
> 
> Looks like the test you added failed in Github Action. https://github.com/jerrypeng/spark/runs/4850411860?check_suite_focus=true
> 
> Could you please look into it?

**Comment by jerrypeng** [comment]
@HeartSaVioR added the fixes

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala [hunk 2]
```diff
@@ -195,6 +195,45 @@ abstract class KafkaMicroBatchSourceSuiteBase extends KafkaSourceSuiteBase {
     true
   }
 
+  test("Trigger.AvailableNow") {
+    val topic = newTopic()
+    testUtils.createTopic(topic, partitions = 5)
+
+    testUtils.sendMessages(topic, (0 until 15).map(x => {
```
Comment (HeartSaVioR) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:202]: nit: we tend to compose ( and { if possible. For this case, you can use `{ case x =>` to get rid of (.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala [hunk 3]
```diff
@@ -195,6 +195,45 @@ abstract class KafkaMicroBatchSourceSuiteBase extends KafkaSourceSuiteBase {
...
+      "foo-" + x
```
Comment (HeartSaVioR) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:203]: nit: We prefer to use string interpolation: s"foo-$x"

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala [hunk 4]
```diff
@@ -195,6 +195,45 @@ abstract class KafkaMicroBatchSourceSuiteBase extends KafkaSourceSuiteBase {
...
+      .option("kafka.metadata.max.age.ms", "1")
+      .option("maxOffsetsPerTrigger", 5)
+      .option("subscribe", topic)
+      .option("startingOffsets", "earliest")
+      .load()
+
+    var index: Int = 0
+    def startTriggerAvailableNowQuery(): StreamingQuery = {
+      reader.writeStream
+        .foreachBatch((df: Dataset[Row], i: Long) => {
```
Comment (HeartSaVioR) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:219]: nit: same reason, please try `{ case (df: Dataset[Row], i: Long) =>` instead and remove (.

**Review by HeartSaVioR — COMMENTED** [review]
Only style guides and looks great!

**Review by HeartSaVioR — COMMENTED** [review]
Only style guides and looks great!

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala [hunk 5]
```diff
@@ -57,7 +57,7 @@ private[kafka010] class KafkaMicroBatchStream(
     metadataPath: String,
     startingOffsets: KafkaOffsetRangeLimit,
     failOnDataLoss: Boolean)
-  extends SupportsAdmissionControl with ReportsSourceMetrics with MicroBatchStream with Logging {
+  extends SupportsTriggerAvailableNow with ReportsSourceMetrics with MicroBatchStream with Logging {
```
Comment (alex-balikov) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala:60]: what about SupportsAdmissionControl - I see other APIs specifying both - https://livegrep.dev.databricks.com/view/databricks/runtime/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala#L57

even if one extends the other, is it better to list them explicitly?

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala [hunk 6]
```diff
@@ -98,7 +100,7 @@ private[kafka010] class KafkaMicroBatchStream(
     } else if (minOffsetPerTrigger.isDefined) {
       ReadLimit.minRows(minOffsetPerTrigger.get, maxTriggerDelayMs)
     } else {
-      maxOffsetsPerTrigger.map(ReadLimit.maxRows).getOrElse(super.getDefaultReadLimit)
+      maxOffsetsPerTrigger.map(ReadLimit.maxRows).getOrElse(ReadLimit.allAvailable())
```
Comment (alex-balikov) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala:103]: why this change? Isn't it better to delegate to the parent if you have no opinion - e.g. the parent could use some config default or whatever.

Comment (HeartSaVioR) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala:103]: Previous comment seems to be out of date. It's to avoid Scala compiler bug.

While we are here, probably good to comment for the reason, and tag `// TODO: `

Comment (HeartSaVioR) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala:103]: https://github.com/apache/spark/pull/35238#discussion_r786502742

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala (hunk 5 above)
Comment (jerrypeng) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala:60]: I don't think it is necessary. It is redundant

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala (hunk 2 above)
Comment (jerrypeng) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:202]: sure I can change this I was just this copied format from existing code

https://github.com/apache/spark/blob/master/external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala#L253

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala (hunk 3 above)
Comment (jerrypeng) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:203]: sure will change

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala (hunk 4 above)
Comment (jerrypeng) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:219]: sure I can change this I just this copied format from existing code

https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala#L1354

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala (hunk 6 above)
Comment (jerrypeng) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala:103]: I will add a comment

**Comment by jerrypeng** [comment]
@HeartSaVioR @alex-balikov thank you for the review.  I have address your comments.  Please take another look.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala [hunk 7]
```diff
@@ -199,9 +199,9 @@ abstract class KafkaMicroBatchSourceSuiteBase extends KafkaSourceSuiteBase {
     val topic = newTopic()
     testUtils.createTopic(topic, partitions = 5)
 
-    testUtils.sendMessages(topic, (0 until 15).map(x => {
-      "foo-" + x
-    }).toArray, Some(0))
+    testUtils.sendMessages(topic, (0 until 15).map{ case x =>
```
Comment (HeartSaVioR) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:202]: sorry for the nits: space between `p` and `{` / same happens in foreachBatch

Not sure style checker will check this, but it is good practice to run`dev/scalastyle` before pushing the change. Not all styles are checked with style checker unfortunately. 

Please go through https://github.com/databricks/scala-style-guide when you have some time.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala [hunk 8]
```diff
@@ -195,6 +195,45 @@ abstract class KafkaMicroBatchSourceSuiteBase extends KafkaSourceSuiteBase {
...
+      .option("kafka.metadata.max.age.ms", "1")
+      .option("maxOffsetsPerTrigger", 5)
+      .option("subscribe", topic)
+      .option("startingOffsets", "earliest")
+      .load()
+
+    var index: Int = 0
+    def startTriggerAvailableNowQuery(): StreamingQuery = {
+      reader.writeStream
+        .foreachBatch{ case (_: Dataset[Row], _: Long) =>
```
Comment (HeartSaVioR) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:219]: My apologize for back and forth. I roughly remember this worked, but looks like it doesn't work. Please revert the latest change.

```
[error] /home/runner/work/spark/spark/external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:219:22: missing parameter type for expanded function
[error] The argument types of an anonymous function must be fully known. (SLS 8.5)
[error] Expected type was: ?
[error]         .foreachBatch{ case (_: Dataset[Row], _: Long) =>
[error]                      ^
[error] one error found
```

**Review by HeartSaVioR — APPROVED** [review]
+1

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala (hunk 6 above)
Comment (xuanyuanking) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala:106]: how about fill a Spark Jira with the detailed description. Then we just need a single line comment here like:
```
TODO(SPARK-xxxx): directly call super.getDefaultReadLimit when ...
```

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala [hunk 9]
```diff
@@ -77,7 +77,9 @@ private[kafka010] class KafkaSource(
     metadataPath: String,
     startingOffsets: KafkaOffsetRangeLimit,
     failOnDataLoss: Boolean)
-  extends SupportsAdmissionControl with Source with Logging {
+  extends SupportsTriggerAvailableNow
+  with Source
+  with Logging {
```
Comment (xuanyuanking) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala:82]: super nit: maybe still keep all the extends in one-line?

**Review by xuanyuanking — APPROVED** [review]
Just some small comments.
LGTM

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala [hunk 10]
```diff
@@ -98,7 +100,10 @@ private[kafka010] class KafkaMicroBatchStream(
     } else if (minOffsetPerTrigger.isDefined) {
       ReadLimit.minRows(minOffsetPerTrigger.get, maxTriggerDelayMs)
     } else {
-      maxOffsetsPerTrigger.map(ReadLimit.maxRows).getOrElse(super.getDefaultReadLimit)
+      // Calling ReadLimit.allAvailable() explicitly instead of super.getDefaultReadLimit because
+      // scala compiler bug https://github.com/scala/bug/issues/12523
+      // TODO revert to calling super.getDefaultReadLimit when scala complier bug is fixed
```
Comment (viirya) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala:105]: It is better to create a JIRA ticket and add into the comment here for tracking it.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala [hunk 11]
```diff
@@ -130,7 +134,10 @@ private[kafka010] class KafkaSource(
     } else if (minOffsetPerTrigger.isDefined) {
       ReadLimit.minRows(minOffsetPerTrigger.get, maxTriggerDelayMs)
     } else {
-      maxOffsetsPerTrigger.map(ReadLimit.maxRows).getOrElse(super.getDefaultReadLimit)
+      // Calling ReadLimit.allAvailable() explicitly instead of super.getDefaultReadLimit because
+      // scala compiler bug https://github.com/scala/bug/issues/12523
+      // TODO revert to calling super.getDefaultReadLimit when scala complier bug is fixed
```
Comment (viirya) [inline external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala:139]: ditto.

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala [hunk 12]
```diff
@@ -195,6 +195,45 @@ abstract class KafkaMicroBatchSourceSuiteBase extends KafkaSourceSuiteBase {
...
+      .option("maxOffsetsPerTrigger", 5)
+      .option("subscribe", topic)
+      .option("startingOffsets", "earliest")
+      .load()
+
+    var index: Int = 0
+    def startTriggerAvailableNowQuery(): StreamingQuery = {
+      reader.writeStream
+        .foreachBatch((_: Dataset[Row], _: Long) => {
+          index+=1
```
Comment (viirya) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:220]: style: `index += 1`

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala [hunk 13]
```diff
@@ -195,6 +195,45 @@ abstract class KafkaMicroBatchSourceSuiteBase extends KafkaSourceSuiteBase {
...
+        })
+        .trigger(Trigger.AvailableNow)
+        .start()
+    }
+
+    val q1 = startTriggerAvailableNowQuery()
```
Comment (viirya) [inline external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:226]: nit: just `query`? there is only one query.

**Review by viirya — APPROVED** [review]
A few minor comments.

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala (hunk 9 above)
Comment (jerrypeng) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala:82]: sure

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala (hunk 13 above)
Comment (jerrypeng) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:226]: sure

File: external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala (hunk 12 above)
Comment (jerrypeng) [reply external/kafka-0-10-sql/src/test/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchSourceSuite.scala:220]: sure

File: external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala (hunk 6 above)
Comment (jerrypeng) [reply external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaMicroBatchStream.scala:106]: sure

**Comment by jerrypeng** [comment]
@viirya and @xuanyuanking thank you for the review.  I have addressed your comments.

**Comment by xuanyuanking** [comment]
Will merge this right after the test passes. Thanks for the contribution and welcome to the community! @jerrypeng

**Comment by HeartSaVioR** [comment]
@xuanyuanking Friendly reminder. Please go with merging at your convenience. Thanks!

**Comment by dongjoon-hyun** [comment]
+1, late LGTM. Thank you all!

**Comment by xuanyuanking** [comment]
Merged to master.
